Implement sort for list of tuples

gmagogsfm opened the pull request:
* Implement tuple sort by traversing contained IValue types and generate a lambda function as comparator for sort.
* Tuple, class objects can now arbitrarily nest within each other and still be sortable

Fixes #43219

gmagogsfm committed:
Implement tuple[] sort

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1153,38 +1152,106 @@ Function* checkSortSchema(const c10::TypePtr& list_element_type) {
+  for (const IValue v : list_ivalues) {
+    auto curr_type = v.toTuple()->type();
+    if (*curr_type != *tuple_type) {
+      why_not << "Different tuple types can not be sorted. ";
+      why_not << "Found different tuple types: ";
+      why_not << *tuple_type;
+      why_not << " and ";
+      why_not << *curr_type << ".";
```
```suggestion
      why_not << "Different tuple types can not be sorted. "
                     << "Found different tuple types: "
                     << *tuple_type
                     << " and "
                     << *curr_type << ".";
```

gmagogsfm replied:
Done.

gmagogsfm replied:
Done.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1153,38 +1152,106 @@ Function* checkSortSchema(const c10::TypePtr& list_element_type) {
+  // Tuple type must contain only sortable types
+  auto tuple_type = list_ivalues.get(0).toTuple()->type();
+  if (!isSortableTupleType(tuple_type)) {
+    why_not << "Contained elements in ";
+    why_not << *tuple_type;
+    why_not << " are not sortable. Only Int, Bool, Float, String, Tensor ";
+    why_not << " or Tuples of aforementionted types can be sorted.";
```
```suggestion
    why_not << "Contained elements in "
                   << *tuple_type
                   << " are not sortable. Only Int, Bool, Float, String, Tensor "
                   << " or Tuples of aforementionted types can be sorted.";
```

gmagogsfm replied:
Done.

SplitInfinity commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +463,52 @@ std::ostream& IValue::repr(
+}
+
+bool operator>(const IValue& a, const IValue& b) {
+  return !(a == b || a < b);
```
I wonder if there is an opportunity for performance improvement by implementing this explicitly instead of in terms of `==` and `<`.

But this proposal comes at the cost of increasing code size. I don't have strong feelings either way, just something to think about.

gmagogsfm replied:
Good point. Done.

SplitInfinity commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +463,52 @@ std::ostream& IValue::repr(
+    case IValue::Tag::Int:
+       return a.toInt() < b.toInt();
+    case IValue::Tag::Bool:
+       return a.toBool() == false && b.toBool() == true;
```
Maybe the way you've written is more readable. I don't feel too strongly about this.

```suggestion
       return !a.toBool() && b.toBool();
```

gmagogsfm replied:
Ack. Showing explicit values is indeed slightly more readable like you said.

SplitInfinity reviewed (commented):
Do you have any insight about the performance implications of this change? The sorting functions you removed were templated and converted the entire `List<IValue>` into a `List<T>` before sorting. The `IValue` comparison functions you added check the type tags of the `IValue` instances and call `IValue::toXXX()` on them while sorting. I think this is an probably an improvement in terms of code size (less template instantiations), but I'm curious about the runtime.

gmagogsfm commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +463,52 @@ std::ostream& IValue::repr(
+  const auto& rhs_vals = rhs.elements();
+
+  // Comparing tuples with different length is undefined.
+  TORCH_INTERNAL_ASSERT(lhs_vals.size() == rhs_vals.size());
```
Done.

gmagogsfm commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +463,52 @@ std::ostream& IValue::repr(
+
+bool operator<(const IValue& a, const IValue& b) {
+  // Must be same type to be compared.
+  TORCH_INTERNAL_ASSERT(a.tag == b.tag);
```
Done.

gmagogsfm commented:
> Do you have any insight about the performance implications of this change? The sorting functions you removed were templated and converted the entire `List<IValue>` into a `List<T>` before sorting. The `IValue` comparison functions you added check the type tags of the `IValue` instances and call `IValue::toXXX()` on them while sorting. I think this is an probably an improvement in terms of code size (less template instantiations), but I'm curious about the runtime.

Good point, I changed basic types' sort kernel back to template-based. Tuple sorting is still dependent on runtime type refining and type checking, so the comparison logic of IValue is still retained.

eellison reviewed (commented):
Looks good, I think we need some small changes to the runtime op. Do we need to add the `>` operator ?

It's a little weird now how we allow sorting of basic types, tuples containing them, and also class types, but we don't allow sorting of tuples containing class types. The downside in the current approach with class types is you would have to look up the `<` function on each comparison, but i think that's probably fine ? I dont know how many models do much class sorting. If that is fine we should merge this implementation with the class impleemntation.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1153,38 +1152,106 @@ Function* checkSortSchema(const c10::TypePtr& list_element_type) {
-              << " must be of Tensors, ints, floats, bools, strs or "
-              << "a User Defined Class that defines the __lt__ compare method"
-              << ", got list of " << list_element_type->repr_str() << "\n";
+              << " must be of Tensors, ints, floats, bools, strs, "
```
Yeah, maybe something like `or Tuples of the aforementioned type`.

gmagogsfm replied:
Done, with the caveat that it isn't exactly true for tuple of classes. Let's discuss in the other comment thread.

SplitInfinity commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -46,6 +46,50 @@ bool operator==(const ivalue::Tuple& lhs, const ivalue::Tuple& rhs) {
+  return false;
+}
+
+bool operator>(const ivalue::Tuple& lhs, const ivalue::Tuple& rhs) {
```
https://en.cppreference.com/w/cpp/algorithm/sort says `std::sort` only needs `operator<`, so yeah, maybe. 😮

gmagogsfm replied:
As mentioned in another comment, std::greater<> needs it for reverse sorting.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -4403,19 +4403,61 @@ def foo():
+
+    def test_string_sort(self):
+        def foo(strs: List[str]):
+            strs.sort
```
```suggestion
            strs.sort()
```

SplitInfinity replied:
More importantly, does it work? If so, why?

gmagogsfm replied:
Lol, it works because both scripted and unscripted version returns the unsorted string list. Fixed, also added a filecheck.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -4403,19 +4403,61 @@ def foo():
-        self.assertEqual(foo(inputs), sorted(inputs))
+        self.checkScript(foo, (inputs,))
+
+    def test_tuple_sorted(self):
```
In light of @eellison's comment, can you add tests for sorting in `reverse` order?

gmagogsfm replied:
Done.

gmagogsfm commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +507,50 @@ std::ostream& IValue::repr(
+  }
+}
+
+bool operator>(const IValue& a, const IValue& b) {
```
Yeah, I am using std::greater<>() to implement reverse sorting, it relies on operator> .

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1153,38 +1152,106 @@ Function* checkSortSchema(const c10::TypePtr& list_element_type) {
-      });
+
+  if (g_list.empty()) {
+    if (copy_return_list) {
```
Done. Reduced to two times, each maps to one distinct action required.

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1153,38 +1152,106 @@ Function* checkSortSchema(const c10::TypePtr& list_element_type) {
+    if (!isSortableListOfTuple(g_list, ss)) {
+      throw std::runtime_error(ss.str());
+    }
+    std::sort(g_list.begin(), g_list.end());
```
Oops, brainfarted. Really shouldn't have done this at midnight.

gmagogsfm reviewed (commented):
> Looks good, I think we need some small changes to the runtime op. Do we need to add the `>` operator ?
> 
> It's a little weird now how we allow sorting of basic types, tuples containing them, and also class types, but we don't allow sorting of tuples containing class types. The downside in the current approach with class types is you would have to look up the `<` function on each comparison, but i think that's probably fine ? I dont know how many models do much class sorting. If that is fine we should merge this implementation with the class impleemntation.

Right, tuple of classes can't be sorted due to how classes are handled (I assume that's what you mean by current approach?). I think the main benefit of looking up "<" function for each class is that objects of different class types can be compared against each other as long as they have __lt__ implemented with right type. Do you mean this is an overhead that we should get rid of?

As for merging "this implementation with class implementation", do you mean we should unify every comparison logic into that of IValue? so they can arbitrarily nest?

gmagogsfm committed:
Change method to generate sorting comparator so that tuple of class objects can be sorted,

eellison commented:
@gmagogsfm could you comment on the large changes ? Why did we have to add the `getLessThanComparator` ?

gmagogsfm commented:
> @gmagogsfm could you comment on the large changes ? Why did we have to add the `getLessThanComparator` ?

This new change is actually still WIP. Sorry for pushing it to gh early, I wanted CI to run. Will update comment once some more changes are done.

eellison commented:
You're good, just trying to do my review requested queue. Removing my name for now you can re-request when its ready.

gmagogsfm commented:
> You're good, just trying to do my review requested queue. Removing my name for now you can re-request when its ready.

Just pushed another revision. The main reason behind this change to to make tuples sortable even if there are other types nested in it. 

`getIValueLessThanComparator` is a one time cost to construct a comparator function specific to each tuple type. Combined with the fact that we enforce all tuple types to be exactly same at runtime check, we can use this comparator to compare every pair of tuples. Since `getIValueLessThanComparator` also supports classes, now class objects can nest inside tuples and be sortable.

Tests are still running but weekend is right ahead of us, so pushing it out to review now for some early comments.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,149 @@ RegisterOperators reg2({
+  }
+
+  // All tuples must have exactly same type.
+  for (const IValue v : list_ivalues) {
```
```suggestion
  for (const IValue& v : list_ivalues) {
```

SplitInfinity commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +464,195 @@ std::ostream& IValue::repr(
+        elements_lts.push_back(getLessThanComparator(elements[i]));
+      }
+
+      return [elements_lts, n](const IValue& a, const IValue& b) {
```
Can we move `elements_lts` into the lambda instead of copying it?

SplitInfinity commented on aten/src/ATen/core/ivalue.cpp:
```diff
@@ -463,6 +464,195 @@ std::ostream& IValue::repr(
+        elements_lts.push_back(getLessThanComparator(elements[i]));
+      }
+
+      return [elements_lts, n](const IValue& a, const IValue& b) {
```
Same here.

SplitInfinity commented on test/test_jit.py:
```diff
@@ -4404,19 +4404,77 @@ def foo():
+        with self.assertRaisesRegexWithHighlight(RuntimeError, "are not sortable", "tups.sort"):
+            foo()
+
+    def test_tuple_unsortable_diff_type(self):
```
Consider adding a test case to test the error message for unsortability due to unsupported types or differing types in a nested tuple. e.g. 
```
[(1, (2, {1: 2})), (1, (2, {1: 2})), (1, (0, {1: 2}))]
```

gmagogsfm replied:
Done. Added a test for different types. Unsortable types is covered in an earlier test.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,154 @@ RegisterOperators reg2({
 void sort_op(Stack* stack) {
   bool reverse = has_reverse_arg ? pop(stack).toBool() : false;
   auto g_list = pop(stack).toList();
+
```
Intentional?

gmagogsfm replied:
yes, slightly clearer.

SplitInfinity commented on test/cpp/tensorexpr/test_reductions.cpp:
```diff
@@ -16,7 +16,6 @@
 #include "torch/csrc/jit/tensorexpr/loopnest.h"
 #include "torch/csrc/jit/tensorexpr/tensor.h"
 
-
```
Accident?

gmagogsfm replied:
Yes, fixed

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -776,7 +777,6 @@ RegisterOperators reg2({
         "aten::sorted.str(str[](a) input) -> (str[])",
         listCopyAndSort<std::string>,
         aliasAnalysisFromSchema()),
-
```
Intentional?

gmagogsfm replied:
Yes, to fix a previous accident lol

SplitInfinity reviewed (commented):
All my comments have been addressed, thanks. I will defer to @eellison for approval.

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,154 @@ RegisterOperators reg2({
 
-bool simpleClassTypeArg(const Argument& arg, const ClassTypePtr& type) {
-  return arg.type() == type && !arg.kwarg_only() && !arg.default_value();
+bool isSortableTupleType(
```
It's surprising to me that we have a separate `isSortableTupleType` - could we refactor this to `isSortableType`, and within the Tuple Case recurse on the type members?

gmagogsfm replied:
done.

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,154 @@ RegisterOperators reg2({
+    return false;
+  }
+
+  // All tuples must have exactly same type.
```
Similarly - this seems like a general restriction of list sorting, would be nice to refactor to remove tuple-specific casing

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,154 @@ RegisterOperators reg2({
+bool isSortableListOfObjects(
+    c10::List<IValue>& objects,
+    std::stringstream& why_not) {
+  auto class_type = objects.get(0).toObject()->type();
```
Could we try to refactor this with isSortableListofTuples?

eellison reviewed (commented):
This looks good, would be nice to refactor tuples and list sorting to reuse more logic

gmagogsfm commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,154 @@ RegisterOperators reg2({
+    }
+  }
+
   if (copy_return_list) {
```
Discussed offline, this is slightly clearer.

eellison commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1137,65 +1137,105 @@ RegisterOperators reg2({
   }
-  throw std::runtime_error(error_str.str());
+
+  if (auto tuple_type = type->cast<TupleType>()) {
```
Can we refactor `isSortableTupleType` to `isSortableType` and then just call that here instead of the two separate branches for Class / Tuple ? It would be a very small change - change input of isSortableTupleType to a single type instead of tuple_type and not iterate and it works

gmagogsfm replied:
This is actually intentional with a very subtle reason. 

If we use `isSortableType` which, by definition, returns true for other basic types (numbers, strings and tensors), that creates the hazard of someone mistakenly messed up schema matching of `aten::sort.basic_types` and we end up missing the chance of throwing an error here that clearly says type is incorrect because now we need to return true for them. Not a huge deal, but would like to know what you think on thisl.

eellison reviewed (approved):
LGTM! Thanks for making the changes. One small remaining nit
